ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/io/BigQuery.java

```diff
@@ -0,0 +1,124 @@
+
+public class BigQuery {
+
+  public static class Write implements Function<Write.TableRow, CompletableFuture<Write.TableRow>> {
+
+    private final com.google.cloud.bigquery.BigQuery bigquery;
+    private final Map<TableId, Batch> batches = new HashMap<>();
+    private final int maxBytes = 10000000; // HTTP request size limit: 10 MB
+    private final int maxMessages = 10000; // Maximum rows per request: 10,000
+    private final long maxDelayMillis = 1000; // default to 1 second
+
+    public Write(com.google.cloud.bigquery.BigQuery bigquery) {
+      this.bigquery = bigquery;
+    }
+
+    @Override
+    public CompletableFuture<TableRow> apply(TableRow row) {
+      return Optional.ofNullable(batches.get(row.tableId)).flatMap(batch -> batch.add(row))
+          .orElseGet(() -> {
+            final Batch batch = new Batch(bigquery, row.tableId);
+            final CompletableFuture<TableRow> response = batch.add(row)
+                .orElseThrow(() -> new IllegalArgumentException("Empty batch rejected message"));
+            batches.put(row.tableId, batch);
+            return response;
+          });
+    }
+
+    /**
+     * POJO of the information necessary to insert a row into BigQuery.
+     */
+    public static class TableRow {
+
+      public final TableId tableId;
+      public final int byteSize;
+      public final Map<String, Object> content;
+
+      public TableRow(TableId tableId, int byteSize, Map<String, Object> content) {
+        this.tableId = tableId;
+        this.byteSize = byteSize;
+        this.content = content;
+      }
+    }
+
+    private class WriteErrors extends RuntimeException {
+
+      public final List<BigQueryError> errors;
+
+      private WriteErrors(List<BigQueryError> errors) {
+        this.errors = errors;
+      }
+    }
+
+    private class Batch {
+
+      private final com.google.cloud.bigquery.BigQuery bigquery;
+      private final InsertAllRequest.Builder builder;
+      private final CompletableFuture<Void> full;
+      private final CompletableFuture<InsertAllResponse> result;
+
+      int size = 0;
+      int byteSize = 0;
+
+      private Batch(com.google.cloud.bigquery.BigQuery bigquery, TableId tableId) {
+        this.bigquery = bigquery;
+        builder = InsertAllRequest.newBuilder(tableId)
+            // ignore row values for columns not present in the table
+            .setIgnoreUnknownValues(true)
+            // insert all valid rows when invalid rows are present in the request
+            .setSkipInvalidRows(true);
+        full = CompletableFuture.runAsync(this::timeout);
+        result = full.handleAsync(this::insertAll);
+      }
+
+      private void timeout() {
+        try {
+          Thread.sleep(maxDelayMillis);
+        } catch (InterruptedException e) {
+          // this is fine
+        }
+      }
+
+      private synchronized InsertAllResponse insertAll(Void v, Throwable t) {
+        return bigquery.insertAll(builder.build());
```

Comment:
It's unclear to me what happens after we call insertAll. I expected that we'd either reset state on this Batch and reuse the object for more rows, or we'd remove this Batch from BigQuery's batches map, so that the next row for this table creates a new batch. But I don't see either of those happening.

Reply:
in `BigQuery.Write.apply` I use `flatMap` to start a new batch when `batch.add(row)` returns `Optional.empty()`:
```
    public CompletableFuture<TableRow> apply(TableRow row) {
      return Optional.ofNullable(batches.get(row.tableId)).flatMap(batch -> batch.add(row))
          .orElseGet(() -> {
            final Batch batch = new Batch(bigquery, row.tableId);
            final CompletableFuture<TableRow> response = batch.add(row)
                .orElseThrow(() -> new IllegalArgumentException("Empty batch rejected message"));
            batches.put(row.tableId, batch);
            return response;
          });
    }
```

Reply:
Interesting. That's subtle enough that I'd recommend making sure the `flatMap` is on its own line, and have a comment above the line to explain that Batch::add will return Optional.empty() as a special flag to indicate it's done and a new batch should be created.

Reply:
it's different now that i'm using `ConcurrentMap`, so it's more obvious